protozfitsreader/__init__.py

```diff
@@ -1,262 +1,313 @@
         else:
-            rawzfitsreader.open("%s:%s" % (self.fname, self.ttype))
+            raise StopIteration
 
-    def _read_message(self):
-        '''Read next message.
-            Fills property self.rawmessage and self.numrows
-        '''
-        self.rawmessage = rawzfitsreader.readEvent()
-        self.numrows = rawzfitsreader.getNumRows()
+    def __iter__(self):
+        return self
 
-    def _extract_field(self, obj, field):
-        # Read a specific field in object 'obj' given as input 'field'
-        if not obj.HasField(field):
-            raise KeyError(str(field))
-        return getattr(obj, field)
+    def __open_events(self):
+        if not self.is_events_table_open:
+            rawzfitsreader.open(self.fname + ":Events")
+            self.is_events_table_open = True
 
     def list_tables(self):
         return rawzfitsreader.listAllTables(self.fname)
 
-    def read_runheader(self):
-        if not self.ttype == "RunHeader":
-            self.ttype = "RunHeader"
-            self._read_file()
-
-        self._read_message()
-        self.header = L0_pb2.CameraRunHeader()
-        self.header.ParseFromString(self.rawmessage)
-
-    def read_event(self):
-        if not self.ttype == "Events":
-            self.ttype = "Events"
-            self._read_file()
-            self.eventnumber = 1
-        else:
-            self.eventnumber += 1
-
-        self._read_message()
-        self.event = L0_pb2.CameraEvent()
-        self.event.ParseFromString(self.rawmessage)
-
     def rewind_table(self):
         # Rewind the current reader. Go to the beginning of the table.
         rawzfitsreader.rewindTable()
 
-    def move_to_next_event(self):
-        # Iterate over events
-        i = 0
-        numrows = i+2
-        # Hook to deal with file with no header (1)
-        if hasattr(self, 'numrows'):
-            numrows = self.numrows
-        # End - Hook to deal with file with no header (1)
-
-        while i < numrows:
-            self.read_event()
-            # Hook to deal with file with no header (2)
-            if hasattr(self, 'numrows'):
-                numrows = self.numrows
-            # End - Hook to deal with file with no header (2)
 
-            # Hook to deal with file with no header (3)
+class Event:
+    def __init__(self, event, run_id, event_id):
+        self.event_id = event_id
+        self.run_id = run_id
+        self._event = event
+
+        _e = self._event                   # just to make lines shorter
+        _w = self._event.hiGain.waveforms  # just to make lines shorter
+
+        self.pixel_ids = toNumPyArray(_w.pixelsIndices)
+        self._sort_ids = np.argsort(self.pixel_ids)
+        self.n_pixels = len(self.pixel_ids)
+        self._samples = toNumPyArray(_w.samples).reshape(self.n_pixels, -1)
+        self.baseline = self.unsorted_baseline[self._sort_ids]
+        self.telescope_id = _e.telescopeID
+        self.event_number = _e.eventNumber
```

Comment:
event_number is a unique identifier fro each event right ? and event_id is starting from 0 incremented by 1?
if so maybe we should switch names ?

Reply:
I am utterly lost with these different `event_number` and `eventNumber` and `event_id` things in SST1M. I believe, this code is functionally identical to the code as it was ~November/December 2017. It is just written down differently. I might have made a mistake though.

Reply:
Yes I found it again: Here is the code as it was in October 25. 
https://github.com/cta-sst-1m/digicampipe/blob/2df1440705328e1a9886018d59959235eb3be454/digicampipe/io/protozfitsreader.py#L255-L257

It had something called "get_event_number" and `self.eventnumber". One was counting up by 1, the other was read from the file. I was puzzled about this duplication then, and I am still puzzled by it. I think everything should be read from the file, the reader should not count events itself.

Reply:
> I think everything should be read from the file, the reader should not count events itself.

I agree if you want to count you can do it outside of the reader. : 

```python
i = 0
for event in ZFits(url):
    i += 1
```